// @powersync/common/lib/client/sync/stream/core-instruction.js
import { FULL_SYNC_PRIORITY } from '../../../db/crud/SyncProgress.js';
function priorityToJs(status) {
    return {
        priority: status.priority,
        hasSynced: status.has_synced ?? undefined,
        lastSyncedAt: status.last_synced_at != null ? new Date(status.last_synced_at * 1000) : undefined
    };
}
export function coreStatusToJs(status) {
    const coreCompleteSync = status.priority_status.find((s) => s.priority == FULL_SYNC_PRIORITY);
    const completeSync = coreCompleteSync != null ? priorityToJs(coreCompleteSync) : null;
    return {
        connected: status.connected,
        connecting: status.connecting,
        dataFlow: {
            // We expose downloading as a boolean field, the core extension reports download information as a nullable
            // download status. When that status is non-null, a download is in progress.
            downloading: status.downloading != null,
            downloadProgress: status.downloading?.buckets,
            internalStreamSubscriptions: status.streams
        },
        lastSyncedAt: completeSync?.lastSyncedAt,
        hasSynced: completeSync?.hasSynced,
        priorityStatusEntries: status.priority_status.map(priorityToJs)
    };
}
