// @powersync/common/lib/client/sync/bucket/OplogEntry.js
import { OpType } from './OpType.js';
export class OplogEntry {
    op_id;
    op;
    checksum;
    subkey;
    object_type;
    object_id;
    data;
    static fromRow(row) {
        return new OplogEntry(row.op_id, OpType.fromJSON(row.op), row.checksum, row.subkey, row.object_type, row.object_id, row.data);
    }
    constructor(op_id, op, checksum, subkey, object_type, object_id, data) {
        this.op_id = op_id;
        this.op = op;
        this.checksum = checksum;
        this.subkey = subkey;
        this.object_type = object_type;
        this.object_id = object_id;
        this.data = data;
    }
    toJSON(fixedKeyEncoding = false) {
        return {
            op_id: this.op_id,
            op: this.op.toJSON(),
            object_type: this.object_type,
            object_id: this.object_id,
            checksum: this.checksum,
            data: this.data,
            // Older versions of the JS SDK used to always JSON.stringify here. That has always been wrong,
            // but we need to migrate gradually to not break existing databases.
            subkey: fixedKeyEncoding ? this.subkey : JSON.stringify(this.subkey)
        };
    }
}
