// @powersync/node/lib/attachments/NodeFileSystemAdapter.js
import { promises as fs } from 'fs';
import * as path from 'path';
import { EncodingType } from '@powersync/common';
/**
 * NodeFileSystemAdapter implements LocalStorageAdapter using Node.js filesystem.
 * Suitable for Node.js environments and Electron applications.
 */
export class NodeFileSystemAdapter {
    storageDirectory;
    constructor(storageDirectory = './user_data') {
        this.storageDirectory = storageDirectory;
    }
    async initialize() {
        const dir = path.resolve(this.storageDirectory);
        await fs.mkdir(dir, { recursive: true });
    }
    async clear() {
        const dir = path.resolve(this.storageDirectory);
        await fs.rmdir(dir, { recursive: true });
    }
    getLocalUri(filename) {
        return path.join(path.resolve(this.storageDirectory), filename);
    }
    async uploadFile(filePath, data, options) {
        const buffer = Buffer.from(data);
        await fs.writeFile(filePath, buffer, {
            encoding: options?.encoding
        });
    }
    async downloadFile(filePath) {
        const data = await fs.readFile(filePath);
        return new Blob([new Uint8Array(data)]);
    }
    async saveFile(filePath, data, options) {
        let buffer;
        if (typeof data === 'string') {
            buffer = Buffer.from(data, options?.encoding ?? EncodingType.Base64);
        }
        else {
            buffer = Buffer.from(data);
        }
        await fs.writeFile(filePath, buffer);
        return buffer.length;
    }
    async readFile(filePath, options) {
        const data = await fs.readFile(filePath);
        if (options?.encoding === EncodingType.Base64) {
            return Buffer.from(data.toString(), 'base64').buffer;
        }
        else {
            return data.buffer.slice(data.byteOffset, data.byteOffset + data.byteLength);
        }
    }
    async deleteFile(path, options) {
        await fs.unlink(path).catch((err) => {
            if (err.code !== 'ENOENT') {
                throw err;
            }
        });
    }
    async fileExists(filePath) {
        try {
            await fs.access(filePath);
            return true;
        }
        catch {
            return false;
        }
    }
    async makeDir(path) {
        await fs.mkdir(path, { recursive: true });
    }
    async rmDir(path) {
        await fs.rmdir(path, { recursive: true });
    }
}
