// @powersync/common/lib/utils/MetaBaseObserver.js
import { BaseObserver } from './BaseObserver.js';
/**
 * A BaseObserver that tracks the counts of listeners for each event type.
 */
export class MetaBaseObserver extends BaseObserver {
    get listenerCounts() {
        const counts = {};
        let total = 0;
        for (const listener of this.listeners) {
            for (const key in listener) {
                if (listener[key]) {
                    counts[key] = (counts[key] ?? 0) + 1;
                    total++;
                }
            }
        }
        return {
            ...counts,
            total
        };
    }
    get listenerMeta() {
        return {
            counts: this.listenerCounts,
            // Allows registering a meta listener that will be notified of changes in listener counts
            registerListener: (listener) => {
                return this.metaListener.registerListener(listener);
            }
        };
    }
    metaListener;
    constructor() {
        super();
        this.metaListener = new BaseObserver();
    }
    registerListener(listener) {
        const dispose = super.registerListener(listener);
        const updatedCount = this.listenerCounts;
        this.metaListener.iterateListeners((l) => {
            l.listenersChanged?.(updatedCount);
        });
        return () => {
            dispose();
            const updatedCount = this.listenerCounts;
            this.metaListener.iterateListeners((l) => {
                l.listenersChanged?.(updatedCount);
            });
        };
    }
}
