// @powersync/common/lib/attachments/AttachmentService.js
import { Mutex } from '../utils/mutex.js';
import { AttachmentContext } from './AttachmentContext.js';
import { AttachmentState } from './Schema.js';
/**
 * Service for querying and watching attachment records in the database.
 *
 * @internal
 */
export class AttachmentService {
    db;
    logger;
    tableName;
    mutex = new Mutex();
    context;
    constructor(db, logger, tableName = 'attachments', archivedCacheLimit = 100) {
        this.db = db;
        this.logger = logger;
        this.tableName = tableName;
        this.context = new AttachmentContext(db, tableName, logger, archivedCacheLimit);
    }
    /**
     * Creates a differential watch query for active attachments requiring synchronization.
     * @returns Watch query that emits changes for queued uploads, downloads, and deletes
     */
    watchActiveAttachments({ throttleMs } = {}) {
        this.logger.info('Watching active attachments...');
        const watch = this.db
            .query({
            sql: /* sql */ `
          SELECT
            *
          FROM
            ${this.tableName}
          WHERE
            state = ?
            OR state = ?
            OR state = ?
          ORDER BY
            timestamp ASC
        `,
            parameters: [AttachmentState.QUEUED_UPLOAD, AttachmentState.QUEUED_DOWNLOAD, AttachmentState.QUEUED_DELETE]
        })
            .differentialWatch({ throttleMs });
        return watch;
    }
    /**
     * Executes a callback with exclusive access to the attachment context.
     */
    async withContext(callback) {
        return this.mutex.runExclusive(async () => {
            return callback(this.context);
        });
    }
}
