// @powersync/common/lib/attachments/Schema.js
import { column } from '../db/schema/Column.js';
import { Table } from '../db/schema/Table.js';
export const ATTACHMENT_TABLE = 'attachments';
/**
 * Maps a database row to an AttachmentRecord.
 *
 * @param row - The database row object
 * @returns The corresponding AttachmentRecord
 *
 * @experimental
 */
export function attachmentFromSql(row) {
    return {
        id: row.id,
        filename: row.filename,
        localUri: row.local_uri,
        size: row.size,
        mediaType: row.media_type,
        timestamp: row.timestamp,
        metaData: row.meta_data,
        hasSynced: row.has_synced === 1,
        state: row.state
    };
}
/**
 * AttachmentState represents the current synchronization state of an attachment.
 *
 * @experimental
 */
export var AttachmentState;
(function (AttachmentState) {
    AttachmentState[AttachmentState["QUEUED_UPLOAD"] = 0] = "QUEUED_UPLOAD";
    AttachmentState[AttachmentState["QUEUED_DOWNLOAD"] = 1] = "QUEUED_DOWNLOAD";
    AttachmentState[AttachmentState["QUEUED_DELETE"] = 2] = "QUEUED_DELETE";
    AttachmentState[AttachmentState["SYNCED"] = 3] = "SYNCED";
    AttachmentState[AttachmentState["ARCHIVED"] = 4] = "ARCHIVED"; // Attachment has been orphaned, i.e. the associated record has been deleted
})(AttachmentState || (AttachmentState = {}));
/**
 * AttachmentTable defines the schema for the attachment queue table.
 *
 * @internal
 */
export class AttachmentTable extends Table {
    constructor(options) {
        super({
            filename: column.text,
            local_uri: column.text,
            timestamp: column.integer,
            size: column.integer,
            media_type: column.text,
            state: column.integer, // Corresponds to AttachmentState
            has_synced: column.integer,
            meta_data: column.text
        }, {
            ...options,
            viewName: options?.viewName ?? ATTACHMENT_TABLE,
            localOnly: true,
            insertOnly: false
        });
    }
}
