// @powersync/node/lib/sync/stream/NodeRemote.js
import * as os from 'node:os';
import { AbstractRemote, DEFAULT_REMOTE_LOGGER, FetchImplementationProvider } from '@powersync/common';
import { BSON } from 'bson';
import { EnvHttpProxyAgent, getGlobalDispatcher, ProxyAgent, WebSocket as UndiciWebSocket } from 'undici';
export const STREAMING_POST_TIMEOUT_MS = 30_000;
class NodeFetchProvider extends FetchImplementationProvider {
    getFetch() {
        return fetch.bind(globalThis);
    }
}
export class NodeRemote extends AbstractRemote {
    connector;
    logger;
    wsDispatcher;
    constructor(connector, logger = DEFAULT_REMOTE_LOGGER, options) {
        const fetchDispatcher = options?.dispatcher ?? defaultFetchDispatcher();
        super(connector, logger, {
            fetchImplementation: options?.fetchImplementation ?? new NodeFetchProvider(),
            fetchOptions: {
                dispatcher: fetchDispatcher
            },
            ...(options ?? {})
        });
        this.connector = connector;
        this.logger = logger;
        this.wsDispatcher = options?.dispatcher;
    }
    createSocket(url) {
        // Create dedicated dispatcher for this WebSocket
        const baseDispatcher = this.getWebsocketDispatcher(url);
        // Create WebSocket with dedicated dispatcher
        const ws = new UndiciWebSocket(url, {
            dispatcher: baseDispatcher,
            headers: {
                'User-Agent': this.getUserAgent()
            }
        });
        return ws;
    }
    getWebsocketDispatcher(url) {
        if (this.wsDispatcher != null) {
            return this.wsDispatcher;
        }
        const protocol = new URL(url).protocol.replace(':', '');
        const proxy = getProxyForProtocol(protocol);
        if (proxy != null) {
            return new ProxyAgent(proxy);
        }
        else {
            return getGlobalDispatcher();
        }
    }
    getUserAgent() {
        return [
            super.getUserAgent(),
            `powersync-node`,
            `node/${process.versions.node}`,
            `${os.platform()}/${os.release()}`
        ].join(' ');
    }
    async getBSON() {
        return BSON;
    }
}
function defaultFetchDispatcher() {
    // EnvHttpProxyAgent automatically uses HTTP_PROXY, HTTPS_PROXY and NO_PROXY env vars by default.
    // We add ALL_PROXY support.
    return new EnvHttpProxyAgent({
        httpProxy: getProxyForProtocol('http'),
        httpsProxy: getProxyForProtocol('https')
    });
}
function getProxyForProtocol(protocol) {
    return (process.env[`${protocol.toLowerCase()}_proxy`] ??
        process.env[`${protocol.toUpperCase()}_PROXY`] ??
        process.env[`all_proxy`] ??
        process.env[`ALL_PROXY`]);
}
