// @powersync/common/lib/db/DBAdapter.js
/**
 * Set of generic interfaces to allow PowerSync compatibility with
 * different SQLite DB implementations.
 */
/**
 * Implements {@link DBGetUtils} on a {@link SqlRunner}.
 */
export function DBGetUtilsDefaultMixin(Base) {
    return class extends Base {
        async getAll(sql, parameters) {
            const res = await this.execute(sql, parameters);
            return res.rows?._array ?? [];
        }
        async getOptional(sql, parameters) {
            const res = await this.execute(sql, parameters);
            return res.rows?.item(0) ?? null;
        }
        async get(sql, parameters) {
            const res = await this.execute(sql, parameters);
            const first = res.rows?.item(0);
            if (!first) {
                throw new Error('Result set is empty');
            }
            return first;
        }
        async executeBatch(query, params = []) {
            // If this context can run batch statements natively, use that.
            // @ts-ignore
            if (super.executeBatch) {
                // @ts-ignore
                return super.executeBatch(query, params);
            }
            // Emulate executeBatch by running statements individually.
            let lastInsertId;
            let rowsAffected = 0;
            for (const set of params) {
                const result = await this.execute(query, set);
                lastInsertId = result.insertId;
                rowsAffected += result.rowsAffected;
            }
            return {
                rowsAffected,
                insertId: lastInsertId
            };
        }
    };
}
/**
 * Update table operation numbers from SQLite
 */
export var RowUpdateType;
(function (RowUpdateType) {
    RowUpdateType[RowUpdateType["SQLITE_INSERT"] = 18] = "SQLITE_INSERT";
    RowUpdateType[RowUpdateType["SQLITE_DELETE"] = 9] = "SQLITE_DELETE";
    RowUpdateType[RowUpdateType["SQLITE_UPDATE"] = 23] = "SQLITE_UPDATE";
})(RowUpdateType || (RowUpdateType = {}));
/**
 * A mixin to implement {@link DBAdapter} by delegating to {@link ConnectionPool.readLock} and
 * {@link ConnectionPool.writeLock}.
 */
export function DBAdapterDefaultMixin(Base) {
    return class extends Base {
        readTransaction(fn, options) {
            return this.readLock((ctx) => TransactionImplementation.runWith(ctx, fn), options);
        }
        writeTransaction(fn, options) {
            return this.writeLock((ctx) => TransactionImplementation.runWith(ctx, fn), options);
        }
        getAll(sql, parameters) {
            return this.readLock((ctx) => ctx.getAll(sql, parameters));
        }
        getOptional(sql, parameters) {
            return this.readLock((ctx) => ctx.getOptional(sql, parameters));
        }
        get(sql, parameters) {
            return this.readLock((ctx) => ctx.get(sql, parameters));
        }
        execute(query, params) {
            return this.writeLock((ctx) => ctx.execute(query, params));
        }
        executeRaw(query, params) {
            return this.writeLock((ctx) => ctx.executeRaw(query, params));
        }
        executeBatch(query, params) {
            return this.writeTransaction((tx) => tx.executeBatch(query, params));
        }
    };
}
class BaseTransaction {
    inner;
    finalized = false;
    constructor(inner) {
        this.inner = inner;
    }
    async commit() {
        if (this.finalized) {
            return { rowsAffected: 0 };
        }
        this.finalized = true;
        return this.inner.execute('COMMIT');
    }
    async rollback() {
        if (this.finalized) {
            return { rowsAffected: 0 };
        }
        this.finalized = true;
        return this.inner.execute('ROLLBACK');
    }
    execute(query, params) {
        return this.inner.execute(query, params);
    }
    executeRaw(query, params) {
        return this.inner.executeRaw(query, params);
    }
    executeBatch(query, params) {
        return this.inner.executeBatch(query, params);
    }
}
class TransactionImplementation extends DBGetUtilsDefaultMixin(BaseTransaction) {
    static async runWith(ctx, fn) {
        let tx = new TransactionImplementation(ctx);
        try {
            await ctx.execute('BEGIN IMMEDIATE');
            const result = await fn(tx);
            await tx.commit();
            return result;
        }
        catch (ex) {
            try {
                await tx.rollback();
            }
            catch (ex2) {
                // In rare cases, a rollback may fail.
                // Safe to ignore.
            }
            throw ex;
        }
    }
}
export function isBatchedUpdateNotification(update) {
    return 'tables' in update;
}
export function extractTableUpdates(update) {
    return isBatchedUpdateNotification(update) ? update.tables : [update.table];
}
