// @powersync/node/lib/db/BetterSqliteWorker.js
import { threadId } from 'node:worker_threads';
class BlockingAsyncDatabase {
    db;
    constructor(db) {
        this.db = db;
        db.function('node_thread_id', () => threadId);
    }
    async close() {
        this.db.close();
    }
    async execute(query, params) {
        const stmt = this.db.prepare(query);
        if (stmt.reader) {
            const rows = stmt.all(params);
            return {
                rowsAffected: 0,
                rows: {
                    _array: rows,
                    length: rows.length
                }
            };
        }
        else {
            const info = stmt.run(params);
            return {
                rowsAffected: info.changes,
                insertId: Number(info.lastInsertRowid)
            };
        }
    }
    async executeRaw(query, params) {
        const stmt = this.db.prepare(query);
        if (stmt.reader) {
            return stmt.raw().all(params);
        }
        else {
            stmt.raw().run(params);
            return [];
        }
    }
    async executeBatch(query, params) {
        params = params ?? [];
        let rowsAffected = 0;
        const stmt = this.db.prepare(query);
        for (const paramSet of params) {
            const info = stmt.run(paramSet);
            rowsAffected += info.changes;
        }
        return { rowsAffected };
    }
}
export async function openDatabase(worker, options) {
    const BetterSQLite3Database = await worker.loadBetterSqlite3();
    const baseDB = new BetterSQLite3Database(options.path);
    baseDB.loadExtension(worker.extensionPath(), 'sqlite3_powersync_init');
    const asyncDb = new BlockingAsyncDatabase(baseDB);
    return asyncDb;
}
