// @powersync/node/lib/db/NodeSqliteWorker.js
import { threadId } from 'node:worker_threads';
import { dynamicImport } from '../utils/modules.js';
class BlockingNodeDatabase {
    db;
    constructor(db, write) {
        this.db = db;
        db.function('node_thread_id', () => threadId);
    }
    async close() {
        this.db.close();
    }
    async execute(query, params) {
        const stmt = this.db.prepare(query);
        const rows = stmt.all(...params);
        return {
            rowsAffected: 0,
            rows: {
                _array: rows,
                length: rows.length
            }
        };
    }
    async executeRaw(query, params) {
        const stmt = this.db.prepare(query);
        stmt.setReturnArrays(true); // Missing in @types/node, https://nodejs.org/api/sqlite.html#statementsetreturnarraysenabled
        return stmt.all(...params);
    }
    async executeBatch(query, params) {
        params = params ?? [];
        let rowsAffected = 0;
        const stmt = this.db.prepare(query);
        for (const paramSet of params) {
            const info = stmt.run(...paramSet);
            rowsAffected += info.changes;
        }
        return { rowsAffected };
    }
}
export async function openDatabase(worker, options) {
    // NOTE: We want to import node:sqlite dynamically, to avoid bundlers unconditionally requiring node:sqlite in the
    // end, since that would make us incompatible with older Node.JS versions.
    const { DatabaseSync } = await dynamicImport('node:sqlite');
    const baseDB = new DatabaseSync(options.path, { allowExtension: true });
    baseDB.loadExtension(worker.extensionPath(), 'sqlite3_powersync_init');
    return new BlockingNodeDatabase(baseDB, options.isWriter);
}
