// @powersync/common/lib/client/CustomQuery.js
import { FalsyComparator } from './watched/processors/comparators.js';
import { DifferentialQueryProcessor } from './watched/processors/DifferentialQueryProcessor.js';
import { OnChangeQueryProcessor } from './watched/processors/OnChangeQueryProcessor.js';
import { DEFAULT_WATCH_QUERY_OPTIONS } from './watched/WatchedQuery.js';
/**
 * @internal
 */
export class CustomQuery {
    options;
    constructor(options) {
        this.options = options;
    }
    resolveOptions(options) {
        return {
            reportFetching: options?.reportFetching ?? DEFAULT_WATCH_QUERY_OPTIONS.reportFetching,
            throttleMs: options?.throttleMs ?? DEFAULT_WATCH_QUERY_OPTIONS.throttleMs,
            triggerOnTables: options?.triggerOnTables
        };
    }
    watch(watchOptions) {
        return new OnChangeQueryProcessor({
            db: this.options.db,
            comparator: watchOptions?.comparator ?? FalsyComparator,
            placeholderData: watchOptions?.placeholderData ?? [],
            watchOptions: {
                ...this.resolveOptions(watchOptions),
                query: this.options.query
            }
        });
    }
    differentialWatch(differentialWatchOptions) {
        return new DifferentialQueryProcessor({
            db: this.options.db,
            rowComparator: differentialWatchOptions?.rowComparator,
            placeholderData: differentialWatchOptions?.placeholderData ?? [],
            watchOptions: {
                ...this.resolveOptions(differentialWatchOptions),
                query: this.options.query
            }
        });
    }
}
